File: src/Data_Structures/Linked_List/linked-list.test.ts
// @ts-nocheck

import { LinkedList } from "./linked-list";

describe("Linked List", () => {
  let list;

  beforeEach(() => {
    list = new LinkedList();
    const values = ["a", "b", "c", "d", "e"];
    values.map((value) => list.push(value));
  });

  test("Should create and manipulate a new Linked list", () => {
    expect(list.isEmpty()).toBeFalsy();
    expect(list.pop().value).toBe("e");
    expect(list.tail.value).toBe("d");
    expect(list.get(1).value).toBe("b");

    list.delete(1);
    expect(list.toArray()).toEqual(["a", "c", "d"]);
    expect(list.pop().value).toBe("d");
    expect(list.toArray()).toEqual(["a", "c"]);

    list.delete(0);
    expect(list.toArray()).toEqual(["c"]);

    list.pop();
    expect(list.toArray()).toEqual([]);

    list.push("f");
    expect(list.toArray()).toEqual(["f"]);

    list.push("g");
    list.push("h");
    expect(list.toArray()).toEqual(["f", "g", "h"]);
    expect(list.shift().value).toBe("f");
    expect(list.toArray()).toEqual(["g", "h"]);

    list.shift();
    expect(list.toArray()).toEqual(["h"]);
    expect(list.unshift("g").value).toBe("g");
    expect(list.toArray()).toEqual(["g", "h"]);

    list.shift();
    expect(list.toArray()).toEqual(["h"]);

    list.shift();
    expect(list.toArray()).toEqual([]);
    expect(list.shift()).toBe(null);
    expect(list.toArray()).toEqual([]);
  });

  test("Should reverse the list", () => {
    expect(list.reverse().toArray()).toEqual(["e", "d", "c", "b", "a"]);
  });

  test("Should find a node in the list", () => {
    expect(list.find("d")).toEqual({
      value: "d",
      next: { value: "e", next: null },
    });
  });
});

File: src/Data_Structures/Linked_List/linked_list_test.py
import pytest

from linked_list import LinkedList


@pytest.fixture(scope="function", autouse=True)
def linked_list():
    linked_list = LinkedList()
    for value in ["a", "b", "c", "d", "e"]:
        linked_list.push(value)

    return linked_list


def test_createAndManipulateLinkedList(linked_list):
    assert linked_list.is_empty() == False
    assert linked_list.pop().value == "e"
    assert linked_list.tail.value == "d"
    assert linked_list.get(1).value == "b"

    linked_list.delete(1)
    assert linked_list.to_list() == ["a", "c", "d"]
    assert linked_list.pop().value == "d"
    assert linked_list.to_list() == ["a", "c"]

    linked_list.delete(0)
    assert linked_list.to_list() == ["c"]

    linked_list.pop()
    assert linked_list.to_list() == []

    linked_list.push("f")
    assert linked_list.to_list() == ["f"]

    linked_list.push("g")
    linked_list.push("h")
    assert linked_list.to_list() == ["f", "g", "h"]
    assert linked_list.shift().value == "f"
    assert linked_list.to_list() == ["g", "h"]

    linked_list.shift()
    assert linked_list.to_list() == ["h"]
    assert linked_list.unshift("g").value == "g"
    assert linked_list.to_list() == ["g", "h"]

    linked_list.shift()
    assert linked_list.to_list() == ["h"]

    linked_list.shift()
    assert linked_list.to_list() == []
    assert linked_list.shift() == None
    assert linked_list.to_list() == []


def test_reverseLinkedList(linked_list):
    assert linked_list.reverse().to_list() == ["e", "d", "c", "b", "a"]


def test_findNodeInLinkedList(linked_list):
    d = linked_list.find("d")
    assert d.value == "d"
    assert d.next.value == "e"
    assert d.next.next == None
